boa/__init__.py

```diff
@@ -37,25 +37,20 @@
 
 @contextlib.contextmanager
 def swap_env(new_env):
-    old_env = env
-    try:
-        set_env(new_env)
+    with set_env(new_env):
         yield
-    finally:
-        set_env(old_env)
 
 
 def set_env(new_env):
     global env
-    env = new_env
-
-    Env._singleton = new_env
+    get_env = lambda: env  # noqa: E731
 
+    def setter(new):
+        global env
+        env = new
+        Env._singleton = new
```

Comment:
i've been using the pattern of moving this to the top-level, and naming it `_set_env`, e.g.
```
def _set_env(new_env):
    global env
    env = new_env
    Env._singleton = new_env
```

Reply:
Sure. Btw whg do we pass a getter to Open instead of the old value?

Reply:
it's more robust, it requires the caller to provide a getter (which is more likely to be correct, since it should be a generic implementation), than forcing the caller to find the correct value